File: experiments-local/experiment-opencode-rerun2/runs/agent=opencode_language=typescript_model=openrouter/z-ai/glm-5.2_tooling=none/rep1/src/routes/books.ts
import { Router } from "express";
import type Database from "better-sqlite3";
import {
  createBook,
  deleteBook,
  getBook,
  listBooks,
  updateBook,
} from "../db.js";
import { formatZodError, parseCreate, parseUpdate } from "../validation.js";
import type { ApiError } from "../types.js";

export function buildBooksRouter(db: Database.Database): Router {
  const router = Router();

  router.get("/", (req, res) => {
    const author = req.query.author;
    const authorFilter = typeof author === "string" ? author : undefined;
    const books = listBooks(db, authorFilter);
    res.json(books);
  });

  router.post("/", (req, res) => {
    let payload;
    try {
      payload = parseCreate(req.body);
    } catch (err) {
      const body: ApiError = formatZodError(err as never);
      return res.status(400).json(body);
    }
    const book = createBook(db, payload);
    res.status(201).json(book);
  });

  router.get("/:id", (req, res) => {
    const id = parseId(req.params.id);
    if (id === null) return res.status(400).json({ error: "invalid_id" });
    const book = getBook(db, id);
    if (!book) return res.status(404).json({ error: "not_found" });
    res.json(book);
  });

  router.put("/:id", (req, res) => {
    const id = parseId(req.params.id);
    if (id === null) return res.status(400).json({ error: "invalid_id" });
    let payload;
    try {
      payload = parseUpdate(req.body);
    } catch (err) {
      const body: ApiError = formatZodError(err as never);
      return res.status(400).json(body);
    }
    const updated = updateBook(db, id, payload);
    if (!updated) return res.status(404).json({ error: "not_found" });
    res.json(updated);
  });

  router.delete("/:id", (req, res) => {
    const id = parseId(req.params.id);
    if (id === null) return res.status(400).json({ error: "invalid_id" });
    const ok = deleteBook(db, id);
    if (!ok) return res.status(404).json({ error: "not_found" });
    res.status(204).send();
  });

  return router;
}

function parseId(raw: string | undefined): number | null {
  if (!raw) return null;
  const n = Number(raw);
  if (!Number.isInteger(n) || n <= 0) return null;
  return n;
}

File: experiments-local/experiment-15/runs/agent=omp_language=rust_model=openrouter/deepseek/deepseek-v3.2_tooling=none/rep2/src/routes/books.rs
use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    routing::{delete, get, post, put},
    Json, Router,
};
use serde::Deserialize;
use uuid::Uuid;
use validator::Validate;

use crate::{
    db::repository::BookRepository,
    error::AppError,
    models::{Book, CreateBookRequest, UpdateBookRequest},
};

#[derive(Debug, Deserialize)]
pub struct BookQuery {
    author: Option<String>,
}

pub fn router(pool: crate::db::DbPool) -> Router {
    let repository = BookRepository::new(pool);
    Router::new()
        .route("/books", post(create_book))
        .route("/books", get(list_books))
        .route("/books/{id}", get(get_book))
        .route("/books/{id}", put(update_book))
        .route("/books/{id}", delete(delete_book))
        .with_state(repository)
}

async fn create_book(
    State(repository): State<BookRepository>,
    Json(payload): Json<CreateBookRequest>,
) -> Result<(StatusCode, Json<Book>), AppError> {
    // Validate input
    payload
        .validate()
        .map_err(|e| AppError::Validation(e.to_string()))?;

    let book = Book::new(payload.title, payload.author, payload.year, payload.isbn);
    let created_book = repository.create(&book).await?;

    Ok((StatusCode::CREATED, Json(created_book)))
}

async fn list_books(
    State(repository): State<BookRepository>,
    Query(query): Query<BookQuery>,
) -> Result<Json<Vec<Book>>, AppError> {
    let books = repository.find_all(query.author).await?;
    Ok(Json(books))
}

async fn get_book(
    State(repository): State<BookRepository>,
    Path(id): Path<Uuid>,
) -> Result<Json<Book>, AppError> {
    let book = repository
        .find_by_id(id)
        .await?
        .ok_or(AppError::NotFound)?;

    Ok(Json(book))
}

async fn update_book(
    State(repository): State<BookRepository>,
    Path(id): Path<Uuid>,
    Json(payload): Json<UpdateBookRequest>,
) -> Result<Json<Book>, AppError> {
    // Validate input if fields are present
    payload
        .validate()
        .map_err(|e| AppError::Validation(e.to_string()))?;

    let existing_book = repository
        .find_by_id(id)
        .await?
        .ok_or(AppError::NotFound)?;

    let updated_book = Book {
        id,
        title: payload.title.unwrap_or(existing_book.title),
        author: payload.author.unwrap_or(existing_book.author),
        year: payload.year.unwrap_or(existing_book.year),
        isbn: payload.isbn.unwrap_or(existing_book.isbn),
    };

    let book = repository.update(id, &updated_book).await?;
    Ok(Json(book))
}

async fn delete_book(
    State(repository): State<BookRepository>,
    Path(id): Path<Uuid>,
) -> Result<(StatusCode, Json<serde_json::Value>), AppError> {
    let deleted = repository.delete(id).await?;
    if deleted {
        Ok((
            StatusCode::NO_CONTENT,
            Json(serde_json::json!({ "message": "Book deleted" })),
        ))
    } else {
        Err(AppError::NotFound)
    }
}